src/completion.ml

```diff
@@ -0,0 +1,147 @@
+
+type rule_ = sym * pp_hint * rule
+
+(** [cp builtins r1 r2] calculates the critical pairs of the rules r1 and r2
+    *)
+let cp : sym StrMap.t -> rule_ -> rule_ -> (term * term) list
+  = fun builtins (s1, h1, r1) (s2, h2, r2) ->
+  let arity1 = Bindlib.mbinder_arity r1.rhs in
+  let arity2 = Bindlib.mbinder_arity r2.rhs in
+  let cps = ref [] in
+  let rec to_m : int -> (meta option) array -> term -> tbox = fun k metas t ->
+    match unfold t with
+    | Vari x         -> _Vari x
+    | Symb (s, h)    -> _Symb s h
+    | Abst (a, t)    ->
+        let (x, t) = Bindlib.unbind t in
+        _Abst (to_m 0 metas a) (Bindlib.bind_var x (to_m 0 metas t))
+    | Appl (t, u)    -> _Appl (to_m (k + 1) metas t) (to_m 0 metas u)
+    | Patt (i, n, a) ->
+        begin
+          let a = Array.map (to_m 0 metas) a in
+          let l = Array.length a in
+          match i with
+          | None   ->
+              let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+              _Meta m a
+          | Some i ->
+              match metas.(i) with
+              | Some m -> _Meta m a
+              | None   ->
+                  let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+                  metas.(i) <- Some m;
+                  _Meta m a
+        end
+    | _              -> assert false in
+  let metas1 = Array.init arity1 (fun _ -> None) in
+  let lhs1 = List.map (fun p -> Bindlib.unbox (to_m 0 metas1 p)) r1.lhs in
+  let lhs1 = Basics.add_args (Symb (s1, h1)) lhs1 in
+  let find_cp : term -> unit = fun t ->
+    let metas1 = Array.init arity1 (fun _ -> None) in
+    let metas2 = Array.init arity2 (fun _ -> None) in
+    let lhs2 = List.map (fun p -> Bindlib.unbox (to_m 0 metas2 p)) r2.lhs in
+    let lhs2 = Basics.add_args (Symb (s2, h2)) lhs2 in
+    let fn m =
+      let m = match m with Some m -> m | None -> assert false in
+      let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
+      let xs = Bindlib.new_mvar mkfree xs in
+      let e = Array.map _Vari xs in
+      TE_Some (Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m e)))
+    in
+    let te_envs1 = Array.map fn metas1 in
+    let te_envs2 = Array.map fn metas2 in
+    let rhs1 = Bindlib.msubst r1.rhs te_envs1 in
+    let rhs2 = Bindlib.msubst r2.rhs te_envs2 in
+    let to_solve = [(t, lhs2)] in
+    match Unif.(solve builtins false {no_problems with to_solve}) with
+    | None    -> ()
+    | Some cs ->
+        if cs <> [] then () else cps := (rhs1, rhs2) :: !cps in
+  Basics.iter find_cp lhs1; (* This only works for algebraic LHS *)
+  !cps
+
+(** [critical_pairs builtins rs] calculates the list of all critical pairs
+    of the rewrite system rs *)
+let rec critical_pairs : sym StrMap.t -> rule_ list -> (term * term) list
+  = fun builtins rs ->
+  match rs with
+  | [] -> []
+  | r :: rs' ->
+      let cps = critical_pairs builtins rs in
+      List.fold_left
+        (fun acc r' ->
+           cp builtins r r' @ cp builtins r' r @ acc)
```

Comment:
It would be more efficient to have cp take an accumulator as well to avoid append's.